### PR TITLE
Add a simple test mode

### PR DESCRIPTION
**Pull request by raphaelm**
- [x] Provide data model and configuration toggle
- [x] Allow to delete individual test orders
- [x] Add tests
- [x] Add a prominent warning message to the backend if test mode orders exist (even though test mode is off), as this leads to wrong statistics
- [x] Decide if and how to generate invoices for test orders as invoice numbers cannot be repeated or should not have gaps.
- [x] Decide if and how we expose test orders through the API, since our difference pull mechanism relies on the fact that orders cannot be deleted.
- [x] Decide if and how we want to couple test modes of payment providers?
- [ ] pretix.eu: Ignore test orders for billing
- [ ] Adjust payment providers: Mollie, bitpay, cash, fakepayment, sepadebit

download

### CODE REVIEW

File: src/pretix/presale/templates/pretixpresale/event/order.html [hunk 1]
```diff
@@ -44,6 +44,9 @@ <h2>
         {% blocktrans trimmed with code=order.code %}
             Your order: {{ code }}
         {% endblocktrans %}
+        {% if order.testmode %}
+            <span class="label label-warning">{% trans "TEST MODE" %}</span>
```
Comment (pc-coholic) [inline src/pretix/presale/templates/pretixpresale/event/order.html]: Minor cosmetic issue: Make `display: inline-block` to match height and baseline of payment-label opposite.
screenshot from 2019-02-19 12-59-33

File: src/pretix/control/templates/pretixcontrol/order/index.html [hunk 2]
```diff
@@ -16,6 +16,9 @@ <h1>
         {% blocktrans trimmed with code=order.code %}
             Order details: {{ code }}
         {% endblocktrans %}
+        {% if order.testmode %}
+            <span class="label label-warning">{% trans "TEST MODE" %}</span>
```
Comment (pc-coholic) [inline src/pretix/control/templates/pretixcontrol/order/index.html]: Minor cosmetic issue: Make `display: inline-block` to match height and baseline of payment-label opposite.

screenshot from 2019-02-19 13-08-20

File: src/pretix/control/templates/pretixcontrol/orders/index.html [hunk 3]
```diff
@@ -117,6 +117,9 @@ <h1>{% trans "Orders" %}</h1>
                                     {{ o.code }}
                                 </a>
```
Comment (pc-coholic) [inline src/pretix/control/templates/pretixcontrol/orders/index.html]: Perhaps we can move the `</a>` to the previous line to get rid of the trailing whitespace which onHover gets converted into a line connecting to the TEST MODE-label.

screenshot from 2019-02-19 13-13-34

File: src/pretix/control/templates/pretixcontrol/search/orders.html [hunk 4]
```diff
@@ -62,6 +62,9 @@ <h1>{% trans "Order search" %}</h1>
                                 {{ o.event.slug|upper }}-{{ o.code }}
                             </a>
```
Comment (pc-coholic) [inline src/pretix/control/templates/pretixcontrol/search/orders.html]: Perhaps we can move the `</a>` to the previous line to get rid of the trailing whitespace which onHover gets converted into a line connecting to the TEST MODE-label.

screenshot from 2019-02-19 13-13-34

File: src/pretix/control/views/event.py [hunk 5]
```diff
@@ -859,6 +860,31 @@ def post(self, request, *args, **kwargs):
...
+                    with transaction.atomic():
+                        for order in request.event.orders.filter(testmode=True):
+                            order.gracefully_delete(user=self.request.user)
+                except ProtectedError:
+                    messages.error(self.request, _('An order could not be deleted as some constraints (e.g. data '
+                                                   'created by plug-ins) do not allow it.'))
+                else:
+                    request.event.cache.set('complain_testmode_orders', False, 30)
+            self.request.event.log_action(
+                'pretix.event.testmode.deactivated', user=self.request.user, data={}
```
[the remark this answers is not in the document]
Comment (raphaelm) [reply src/pretix/control/views/event.py]: Hmm, yes, although .gracefully_delete() creates a log entry for every order

File: src/pretix/presale/templates/pretixpresale/event/checkout_payment.html [hunk 6]
```diff
@@ -31,6 +31,20 @@ <h4 class="panel-title">
                     <div id="payment_{{ p.provider.identifier }}"
                             class="panel-collapse collapsed {% if selected == p.provider.identifier %}in{% endif %}">
                         <div class="panel-body form-horizontal">
+                            {% if request.event.testmode %}
+                                {% if p.provider.test_mode_message %}
+                                    <div class="alert alert-info">
+                                        {{ p.provider.test_mode_message }}
+                                    </div>
+                                {% else %}
+                                    <div class="alert alert-danger">
```
[the remark this answers is not in the document]
Comment (raphaelm) [reply src/pretix/presale/templates/pretixpresale/event/checkout_payment.html]: Heh, that's on purpose so users catch error messages m) Let's just use alert-warning instead?

**Pull request merged**